### ISSUE TITLE
Question about causal training

### ISSUE THREAD

**Issue by kafan1986** [chomeyama/wavehax#5]
You have mentioned causal. But I don't see any causal training implementation. You are generating using overlap and lookahead and throwing away those generated frames? My understanding is correct? Any particular reason why you did not evaluate true causal training which would have not required unnecessary compute during inference?

**Comment by chomeyama** [comment]
Thank you for your question and for carefully looking into the implementation.

At the moment, we have not released the causal version of the model. Due to a recent change in my affiliation and internal constraints, I am not able to share additional code at this time, if requested.

That said, implementing the causal variant itself is fairly straightforward. It can be achieved by making all convolution layers causal, as well as applying causal padding in the STFT computation.

Regarding the lookahead behavior, this is handled by padding the input features (e.g., mel-spectrogram, F0) and then adjusting the effective sequence length in the first Conv1d layers.

Also, the training pipeline does not require special modifications. The same training code used for the non-causal model can be applied as-is.

I hope this clarifies your concerns.

**Comment by kafan1986** [comment]
Thanks for quick reply. One more question, have you done any research related to multi-speaker TTS using MS-wavehax ? Channel 64 and Blocks 8 are enough for multi speaker setup?

**Comment by chomeyama** [comment]
Thanks for your question.

A multi-speaker setup with the setting can work in practice. However, I would recommend fine-tuning the model using features predicted by the acoustic model, or training it in an end-to-end manner.

This is mainly because the model has a relatively small number of parameters, so the quality of the input features has a significant impact on the final output quality.

**Comment by kafan1986** [comment]
@chomeyama I have trained a TTS decoder causal model with the default Channel 64 and Blocks 8 with using features flowing from upstream features. The upstream is based on StyleTTS2 architecture, thus is feature reach in terms of prosody and text encoder features. The modifications made to Decoder to make it causal and use upstream parameters for training has increased parameters size to around 0.55M

The final model was trained on dataset that contained audios belonging to a single speaker (although had variety in terms of  quite emotive expressiveness across samples). Overall dataset size was 25+ hours. 

The model currently suffers from a husky noisy background presence.

The MPD period was set to [2,3,5,7,11] as per original Hifigan implementation, later I saw your config is [2,3,5,7,11,13,17] for Univnet's corresponding MRD module. Not sure how would your config impact this. I have yet to re-run the training with this config. 

I have updated the channels to 128 (from 64), num_blocks to 12 (from 8) and n_fft to 256 (from 128). The target device has constrained processing capability. Any change that you think is overkill, I can reduce further.

Wanted to know how else can I improve quality further? Below is the config I am planning to train. Any suggestions?

Discriminator MRD with  period set to [2,3,5,7,11,13,17] 

        type: 'ms_wavehax'
        channels: 128              # Hidden channels in ConvNeXt blocks (C)
        mult_channels: 2          # Expansion factor in pointwise convs (C -> C*mult -> C)
        kernel_size: 7            # Depthwise conv kernel size (7x7)
        num_blocks: 12            # Number of ConvNeXt blocks
        num_splits: 4             # Number of sub-bands for multi-stream decomposition
        n_fft: 256                # Sub-band STFT FFT size (freq_bins = 65)
        filter_taps: 62           # Analysis/synthesis filter taps
        asr_proj_dim: 128         # ASR feature projection dimension
        sample_rate: 22050        # Must match dataset
        full_hop_length: 256      # Must match preprocess_params.hop_length
        norm_type: 'batch_norm'   # 'batch_norm' (streaming) or 'group_norm' (per-sample)

**Comment by chomeyama** [comment]
Are you using any lookahead?
If not, I’d recommend using at least 1 frame of lookahead; it would help reduce the husky noise and improve speech quality more effectively than just increasing the model size.

**Comment by kafan1986** [comment]
No I am not using any lookahead during training as it is causal. During inference I am only using previous frame only for vocoder (stft/istft operations so as to maintain stft overlap properly) step and not in the CNN part. I am not using a 1 right frame lookahead during generation. If my chunk size is 96 (~ 1 second of audio) then what you are saying I should generate 96+1 right frame and return the first 96 frame corresponding audio data and throw away the right most frame data. This should improve the quality. I think this might improve the boundary artifacts (clicks) issue too that I am currently getting sometimes at the edge between audio chunks. What are your thoughts?

**Comment by chomeyama** [comment]
Yes, your understanding is correct. This would effectively reduce the noise.

**Comment by kafan1986** [comment]
Does increasing n_fft to 256 from current 128 would help? As there will be significant increase in compute required. If it is not worth the extra processing maybe I will use default 128 or something like an intermediate value like 192 or 160.

 n_fft: 256                # Sub-band STFT FFT size (freq_bins = 65)

Any thoughts?

**Comment by chomeyama** [comment]
In most cases, setting n_fft to around 20 ms with a frame shift of 10 ms provides a good balance between quality and computational cost. For example, if you split the signal into four sub-bands, the effective n_fft per band becomes about 5 ms.

Increasing n_fft to 40 ms generally does not justify the additional computation, as the improvement in quality is limited.

**Comment by kafan1986** [comment]
@chomeyama What value do you suggest for mult_channels. I have used 2, would increasing it to 4 or 3 improve quality ?

**Comment by chomeyama** [comment]
@kafan1986 
I’ve tried increasing mult_channels before, and I did see slight improvements in objective metrics such as PESQ, but the impact was not very large. So moving from 2 to 3 or 4 may help a bit, especially if the extra computation is still within your budget, but I wouldn’t expect a major quality gain from it alone. If you have enough compute/memory headroom, it is probably worth trying mult_channels=3 or 4. However, based on my experience, increasing the base channels tends to have a stronger effect on quality than increasing mult_channels. So if you have to choose where to spend the extra compute, I would prioritize slightly larger channels first.
